Simplify and Modernize AJAX Lesson (#1261)
Techtonica/curriculum: ajax/ajax.md

# AJAX

### Projected Time
About 4 hours

- Lesson: 50 mins
- Hands-on Explanation: 50 mins 
- Guided Practice: 1 hr
- Independent Practice: 1 hr
- Check for understanding: 20-30 mins

### Prerequisites

> - [JavaScript Lessons](/javascript)

Reviewer: can you also link to the javascript promises outline?

### Prerequisites
- [JavaScript Lessons](/javascript)
- [jQuery Lesson](/jquery/jquery)

> - [API/JSON Lesson](/api/apis-and-json)

Reviewer 2: Why did you removed the extension ?

Author: The file works without the extension. Should I add it back ?

Reviewer 2: Yeah, add them back.

### Motivation
Front-end code of interactive websites often needs to talk to backend servers to get and present data that the user asks for. Usually, this is done using API calls. AJAX is a way to make asynchronous calls to the server using JavaScript.

### Objectives

**Participants will be able to:**

> - Make AJAX calls to an API to servers.

Reviewer: ```suggestion
- Make AJAX calls to API servers.
```

### Motivation
Front-end code of interactive websites often needs to talk to backend servers to get and present data that the user asks for. Usually, this is done using API calls. AJAX is a way to make asynchronous calls to the server using JavaScript.

### Objectives

**Participants will be able to:**

> - AJAX calls to an API to servers.

Reviewer 2: The line could be rephrased as `Make AJAX calls to an API to servers.`

Author: Okay , will work on it.

### Materials

> - [AJAX Slides](https://drive.google.com/file/d/12KSxnLTCCRl2wG3G3cPjv_9ZF-mltcKx/view?usp=sharing)

Reviewer 2: @alodahl I think the changes should be made in the old presentation only instead of sharing a new file link.

Author: The changes are made to previous slides only. Actually the link provided are restricted to read-only. That is why a new link is provided , with possible changes.

Reviewer 2: You could have requested for edit permission.

Reviewer: Thanks Kundan.  The slides links are stored for apprentices in more places than just here in the outline, so keeping the same link is important so we don't go out of sync!

- [AJAX tutorial on Codecademy.](https://www.codecademy.com/courses/introduction-to-javascript/lessons/requests-i/exercises/requests-intro-i)

> ### Lesson

Reviewer 2: @ashishnagpal2498 What happened in this video ?

Author: The video is preview of old slides only.

Reviewer 2: So I think a new issue for that could be made. @alodahl What do you think ?

Reviewer: @ashishnagpal2498 is going to change the original slides, so we should keep this link.  How  much wlll the slides change? If it's just a little, we can keep this video for a while.

> - Read through lesson slides [AJAX](https://drive.google.com/file/d/12KSxnLTCCRl2wG3G3cPjv_9ZF-mltcKx/view?usp=sharing)

Reviewer 2: @alodahl I think the changes should be made in the old presentation only instead of sharing a new file link.

> Lets say there is an input box and a button. Onclick of button the page shows some data say *hello world* send by the *server*. When the button is clicked -  
#### Full load Application Flow
1. The browser sends a request to a server designated by the URL. 
2. The server does some processing, probably queries a database, then responds to the browser by sending back HTML and JavaScript
3. The browser renders that HTML and JavaScript into a webpage.

#### AJAX Flow
- The client sends a request to the server.
- The server does some processing and returns the corresponding data (usually in JSON format).
- When the request gets completed the callback function is executed in which data is present. That data is shown at frontend in any way you want.

| Difference Point        | Full Load Page          | AJAX  |
| ------------- |:-------------| -----:|
| Client requests to server   | waits for response  | can perform other task simultaneously  |
| Server Response       | responds a corresponding HTML page  |  responds with data (preferably JSON ) |
| Render |  a complete new page is rendered   | only that part of page changes which has requested |
| Thread Blocked | true   | false |

> | Data Received by | the browser in form of HTML   | Javascript(callback or Promise) |

Reviewer 2: Make the change as below.
Javascript -> JavaScript.

Author: Okay

For better understanding follow this [Link](https://github.com/ashishnagpal2498/AjaxVsFullLoadPage)

### What does AJAX do?

> Let's think about how web applications work. Generally, the steps go like this:

Reviewer 2: In line 63 you are talking about steps, then why did you removed the steps ?

Let's say this website that the end user is visiting is an e-commerce site. The end user is browsing a list of products, sees a product they want to buy, and clicks the "Buy" button. This action (clicking the "Buy" button) triggers the request/response steps listed above, this time for the purpose of putting the product in the shopping cart.

Before AJAX, the request/response cycle could only happen with the web page *as a whole*. The end user, by clicking the "Buy" button, would cause the backend server to generate a completely new set of HTML and JavaScript, and the browser would reload and render it as an entirely new page.

AJAX, however, enabled the browser to make requests and only re-render *parts* of a webpage. This was beneficial performance-wise, because smaller strings of HTML, rather than all the HTML for the entire page, was all that was needed to be sent by the server and rendered by the browser. This also made the end user's browsing experience smoother: interactions with the webpage resulted in changes on screen without being interrupted by a reload.

### Remember

> [Thread of Execution](../javascript/javascript-9.md)

Reviewer 2: Make `javascript-9.md` to `javascript-9-async.md`

Author: No such file exist with that path.

- Video walkthrough of lesson slides [AJAX](https://drive.google.com/file/d/1pLyMqvC-8cFroVTWMUBeG1_BRd4L-rL3/view?usp=sharing)

### What is AJAX ?
AJAX stands for Asynchronous JavaScript and XML. It is a **technique** by which front-end get data from the backend by making *asynchronous* API calls.
Because AJAX is *asynchronous*, you must pass in a callback function to handle the received data. Making an AJAX request without any callback function will not have any effect on your application.

> ### Full Load Page V/s AJAX

Reviewer: ```suggestion
### Full Page Load vs AJAX
```

> > Lets say there is an input box and a button. Onclick of button the page shows some data say *hello world* send by the *server*. When the button is clicked -  

Reviewer: ```suggestion
> Let's say there is a text input and a button. On button click, the text "Hello World" is fetched from an API and appears on the page.  Here is what happens when the button is clicked:  
```

#### Full load Application Flow
1. The browser sends a request to a server designated by the URL. 
2. The server does some processing, probably queries a database, then responds to the browser by sending back HTML and JavaScript
3. The browser renders that HTML and JavaScript into a webpage.

#### AJAX Flow
1. The client sends a request to the server.
2. The server does some processing and returns the corresponding data (usually in JSON format).

> 3. When the request gets completed the callback function is executed in which data is received. That data is shown on frontend through JavaScript in any format you want.

Reviewer: ```suggestion
3. When the request gets completed, the callback function is executed with the new data. The callback function could do anything you like; for example, it could use JS to add it to the HTML page.
```

| Difference Point        | Full Load Page          | AJAX  |

> | ------------- |:-----------------:| :---------:|

Reviewer: love the table!

### What does AJAX do?
Let's say this website that the end user is visiting is an e-commerce site. The end user is browsing a list of products, sees a product they want to buy, and clicks the "Buy" button. This action (clicking the "Buy" button) triggers the request/response steps listed above, this time for the purpose of putting the product in the shopping cart.

Before AJAX, the request/response cycle could only happen with the web page *as a whole*. The end user, by clicking the "Buy" button, would cause the backend server to generate a completely new set of HTML and JavaScript, and the browser would reload and render it as an entirely new page.

AJAX, however, enabled the browser to make requests and only re-render *parts* of a webpage. This was beneficial performance-wise, because smaller strings of HTML, rather than all the HTML for the entire page, was all that was needed to be sent by the server and rendered by the browser. This also made the end user's browsing experience smoother: interactions with the webpage resulted in changes on screen without being interrupted by a reload.

### Remember

> [Thread of Execution](/javascript/javascript-async-9.md)

Reviewer: ```suggestion
[Thread of Execution](/javascript/javascript-9-async.md)
```

Reviewer: it doesnt link

### Hands-On Practice 

> * HTML

Reviewer: Be explicit, telling them to create a directory, and what to name the html, css, and JS files.  Also, do they really need a css file, or can they just add this one style inline? As this is the AJAX outline, theres no real need to focus on style here.

> 	* Create a button on which when clicked an API is called and data is fetched.

Reviewer: ```suggestion
	* First: create a button which, when clicked, calls an API to fetch data.
```

```
<button class="btn" onclick="getData()"> Show Data </button>
```
* CSS
	* Add simple styles to the button.	
```
.btn{
	padding: 10px;

}
```
* JS

>   * A function that fetch data from API and displays on the HTML page.

Reviewer: ```suggestion
  * Next: Let's write a function that fetches data from an API and displays it on the HTML page.
```

```
function getData() {
  console.log("Function getData is executed");
  fetch("https://jsonplaceholder.typicode.com/todos/1", {
    headers: { "Content-Type": "application/json" }
  })
    .then(res => res.json())
    .then(data => {
      console.log("data", data);
      let element = document.getElementById("data");
      element.textContent = data.title;
    });
}
```

> 

Reviewer: Don't forget to keep this idea... maybe move to "What is AJAX?"

> [CodePen](https://codepen.io/ashish24_nagpal/pen/NWqXjWN)

Reviewer: Describe what they will find here and what they should do with it.

AJAX, however, enabled the browser to make requests and only re-render *parts* of a webpage. This was beneficial performance-wise, because smaller strings of HTML, rather than all the HTML for the entire page, was all that was needed to be sent by the server and rendered by the browser. This also made the end user's browsing experience smoother: interactions with the webpage resulted in changes on screen without being interrupted by a reload.

### Remember

> [Thread of Execution](/javascript/javascript-async-9)

Reviewer 2: Add file extension.

Author: Okay

Author: what about the slides access ?

Reviewer 2: Yeah, add them back. As confusion would arise.

Reviewer 2: @alodahl Please give him slide access.

- Read [this very short article on Axios](https://flaviocopes.com/axios/), which is an npm package that uses ajax under the hood. Developers like it because it is supported by all legacy and current browsers, and has methods that make some tasks like interpreting received data or setting request timeouts really simple.
- Build an app that gets its data from the Pokémon API!: https://atom-morgan.github.io/ajax/

### Points to Remember
- AJAX is not a programming language. It is a technique.
- It works Asynchronously to fetch data from API.

> - Since it works Asynchronously , data is received in form of Promise or Callback function.

Reviewer 2: Remove whitespace after `Asynchronously`
